File: backend/src/models/property.model.js
const mongoose = require('mongoose');

const propertySchema = new mongoose.Schema({
  proprietaire: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  titre: {
    type: String,
    required: [true, 'Titre est requis'],
    trim: true
  },
  description: {
    type: String,
    required: [true, 'Description est requise']
  },
  type: {
    type: String,
    enum: ['appartement', 'maison', 'studio', 'villa', 'bureau', 'commerce'],
    required: true
  },
  statut: {
    type: String,
    enum: ['disponible', 'loué', 'maintenance', 'inactif'],
    default: 'disponible'
  },
  adresse: {
    rue: {
      type: String,
      required: true
    },
    ville: {
      type: String,
      required: true
    },
    codePostal: {
      type: String,
      required: true
    },
    pays: {
      type: String,
      required: true
    },
    coordonnees: {
      latitude: Number,
      longitude: Number
    }
  },
  caracteristiques: {
    surface: {
      type: Number,
      required: true
    },
    pieces: {
      type: Number,
      required: true
    },
    chambres: Number,
    sallesDeBain: Number,
    etage: Number,
    ascenseur: Boolean,
    parking: Boolean,
    meuble: Boolean,
    jardin: Boolean,
    balcon: Boolean,
    climatisation: Boolean,
    chauffage: String
  },
  loyer: {
    montant: {
      type: Number,
      required: true
    },
    charges: {
      type: Number,
      default: 0
    },
    garantie: Number,
    periodicite: {
      type: String,
      enum: ['mensuel', 'trimestriel', 'annuel'],
      default: 'mensuel'
    }
  },
  photos: [{
    url: String,
    principale: Boolean
  }],
  documents: [{
    type: {
      type: String,
      enum: ['diagnostic', 'assurance', 'taxe', 'autre']
    },
    nom: String,
    url: String,
    dateExpiration: Date
  }],
  disponibilite: {
    dateDisponibilite: Date,
    dureeMinimale: Number, // en mois
    dureeMaximale: Number // en mois
  },
  conditions: {
    animauxAutorises: Boolean,
    fumeurAutorise: Boolean,
    garantRequise: Boolean,
    conditionsParticulieres: String
  },
  evaluations: [{
    locataire: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    note: {
      type: Number,
      min: 1,
      max: 5
    },
    commentaire: String,
    date: {
      type: Date,
      default: Date.now
    }
  }],
  visites: [{
    date: Date,
    visiteur: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    statut: {
      type: String,
      enum: ['demandée', 'confirmée', 'annulée', 'effectuée'],
      default: 'demandée'
    }
  }],
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

// Index pour la recherche géographique
propertySchema.index({ 'adresse.coordonnees': '2dsphere' });

// Index pour la recherche textuelle
propertySchema.index({
  titre: 'text',
  description: 'text',
  'adresse.ville': 'text'
});

module.exports = mongoose.model('Property', propertySchema);

File: locapro_fixed/lib/features/proprietaire/data/models/property_model.dart
import 'property_sub_models.dart';

class PropertyModel {
  final String id;
  final String proprietaireId;
  final String titre;
  final String description;
  final String type;
  final String statut;
  final Address adresse;
  final Characteristics caracteristiques;
  final Rental loyer;
  final List<Photo> photos;
  final List<Document> documents;
  final DateTime? dateDisponibilite;
  final DateTime createdAt;
  final DateTime updatedAt;

  static const List<String> typeOptions = [
    'appartement',
    'maison',
    'studio',
    'villa',
    'bureau',
    'commerce'
  ];

  static const List<String> statutOptions = [
    'disponible',
    'loué',
    'maintenance',
    'inactif'
  ];

  PropertyModel({
    required this.id,
    required this.proprietaireId,
    required this.titre,
    required this.description,
    required this.type,
    required this.statut,
    required this.adresse,
    required this.caracteristiques,
    required this.loyer,
    required this.photos,
    required this.documents,
    this.dateDisponibilite,
    required this.createdAt,
    required this.updatedAt,
  });

  factory PropertyModel.fromJson(Map<String, dynamic> json) {
    return PropertyModel(
      id: json['_id'] ?? '',
      proprietaireId: json['proprietaire'] ?? '',
      titre: json['titre'] ?? '',
      description: json['description'] ?? '',
      type: json['type'] ?? 'appartement',
      statut: json['statut'] ?? 'disponible',
      adresse: Address.fromJson(json['adresse'] ?? {}),
      caracteristiques: Characteristics.fromJson(json['caracteristiques'] ?? {}),
      loyer: Rental.fromJson(json['loyer'] ?? {}),
      photos: (json['photos'] as List<dynamic>? ?? [])
          .map((photo) => Photo.fromJson(photo))
          .toList(),
      documents: (json['documents'] as List<dynamic>? ?? [])
          .map((doc) => Document.fromJson(doc))
          .toList(),
      dateDisponibilite: json['disponibilite']?['dateDisponibilite'] != null
          ? DateTime.parse(json['disponibilite']['dateDisponibilite'])
          : null,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'proprietaire': proprietaireId,
      'titre': titre,
      'description': description,
      'type': type,
      'statut': statut,
      'adresse': adresse.toJson(),
      'caracteristiques': caracteristiques.toJson(),
      'loyer': loyer.toJson(),
      'photos': photos.map((photo) => photo.toJson()).toList(),
      'documents': documents.map((doc) => doc.toJson()).toList(),
      'disponibilite': {
        'dateDisponibilite': dateDisponibilite?.toIso8601String(),
      },
    };
  }

  PropertyModel copyWith({
    String? id,
    String? proprietaireId,
    String? titre,
    String? description,
    String? type,
    String? statut,
    Address? adresse,
    Characteristics? caracteristiques,
    Rental? loyer,
    List<Photo>? photos,
    List<Document>? documents,
    DateTime? dateDisponibilite,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return PropertyModel(
      id: id ?? this.id,
      proprietaireId: proprietaireId ?? this.proprietaireId,
      titre: titre ?? this.titre,
      description: description ?? this.description,
      type: type ?? this.type,
      statut: statut ?? this.statut,
      adresse: adresse ?? this.adresse,
      caracteristiques: caracteristiques ?? this.caracteristiques,
      loyer: loyer ?? this.loyer,
      photos: photos ?? this.photos,
      documents: documents ?? this.documents,
      dateDisponibilite: dateDisponibilite ?? this.dateDisponibilite,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}